# 异常
 
* 异常是有时候可以，有时候不可以
 
* 现在只要触发了异常，程序将终止
 
* 当我们在做一些网络编程的时候，网络情况不一定什么时候是什么样子
 
  ```
  www.baidu.com
  ```
 
* 异常捕获 可以在错误发生时，捕捉到，默认的Python的行为将不执行
 
**IndentationError**: unexpected indent
 
​   缩进错误
 
**ZeroDivisionError**: division by zero
 
​   除数为0
 
**NameError**: name 'b' is not defined
 
​   访问未声明变量
 
**IndexError**: list index out of range
 
​   访问越界(超过了原有数据的长度)变量
 
**KeyError**: 2
 
​   访问字典中(JSON类似)不存在key值
 
**AttributeError**: module 'socket' has no attribute 'create'
 
​   访问一个对象(变量，函数，类，模块)不存在的属性
 

```powershell
try:
    你觉得哪个代码可能出现错误
except TypeError as e: #我要捕获的就是TypeError
    捕获后你要做的事情
    e是python自带的报错信息，如果你需要的话，可以保留他
```
 
* 异常捕获可以提高代码的健壮性
* 让我们的代码在不同情况下，可以让**程序继续向下**，而不是直接中断
* 有一些异常错误，需要额外导入模块才可以使用，切记。
 
```python
try:
    捕获异常
except: 
    捕获后做的事情
else:
    当错误没有发生，那么执行这里的代码
```
 
```python
try:
    print('打开文件')
    fp = open('1.txt')
except FileNotFoundError:
    print('这个文件不存在')
else:
    print("关闭文件")
    fp.close()
#else 分支可以在没有异常出现的时候 执行
```
 
```python
try:
    捕获异常
except:
    捕获要做的事情
finally:
    不管错误是否发生，都执行这里的代码
    做一些收尾工作
```
 
 
 
## 异常的类型
 
```
SystemExit(系统中断异常)
KeyboardInterrupt(ctrl+c)
Exception(内建异常类) 
    Python中的异常都是继承自这个Exception而来的
```

## 手动抛出异常
 
* 手动抛出异常 在测试时候用的多
```powershell
try:
    raise TypeError('我心情不好，我就是要报错')
    #raise语句去将一个合法的异常
except TypeError:
    print('这是我刚才自己要抛出来的异常，现在我要捕获他')
print("哈哈哈哈哈哈")
```
 
```powershell
def func(name,age):
    #name str
    #age int
    if type(name) == str and type(age) == int:
        print('%s:%d' % (name,age,abc))
    else:
        raise TypeError('你能不能好好传参')
    #%d只能接受数字
name = '小明'
age = '18'
func(name,age)
```
 
## 自定义异常
 
* 所有的异常都来自于 **Exception**
 
```python
class MyError(Exception): #面向对象的继承方法
    pass #什么都不干，过！
 
#继承自己的异常基类，从这个基础来创建自己的异常
# 现在已经创建出来了一个自己的错误
 
def func(name,age):
    #name str
    #age int
    if type(name) == str and type(age) == int:
        print('%s:%d' % (name,age,abc))
    else:
        raise MyError('你能不能好好传参')
        #遇到特定情况，手动抛出异常
 
name = '小明'
age = '18'
func(name,age)
```

